File: app/src/main/java/Model/MFBLocation.java
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2021 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package Model;

import android.Manifest;
import android.content.BroadcastReceiver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.database.sqlite.SQLiteDatabase;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.Log;

import com.myflightbook.android.ActNewFlight;
import com.myflightbook.android.MFBMain;
import com.myflightbook.android.R;
import com.myflightbook.android.mfblocationservice;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

public class MFBLocation implements LocationListener {
    public enum GPSQuality {Unknown, Poor, Good, Excellent}

    interface FlightEvents {
        void TakeoffDetected(Location l, Boolean fIsNight);

        void LandingDetected(Location l);

        void FSLandingDetected(Boolean fIsNight);

        void AddNightTime(double t);

        void UpdateStatus(GPSQuality quality, Boolean fAirborne, Location loc, Boolean fRecording);

        boolean shouldKeepListening();
    }

    private Boolean m_fIsEnabled = true;
    private FlightEvents m_Listener = null;

    private static Location m_lastSeenLoc = null; // last location we've seen
    private Location m_currentLoc = null; // last good location we've seen
    private int cSamplesSinceWaking = 0;
    private Boolean fPreviousLocWasNight = false;
    private Location PreviousLoc = null;

    public enum AutoFillOptions {None, FlightTime, EngineTime, HobbsTime, BlockTime, FlightStartToEngineEnd}

    public enum NightCriteria { EndOfCivilTwilight, Sunset, SunsetPlus15, SunsetPlus30, SunsetPlus60 }
    public enum NightLandingCriteria { SunsetPlus60, Night}

    static public Boolean fPrefRecordFlight = false;
    static public Boolean fPrefRecordFlightHighRes = false;
    static public Boolean fPrefAutoDetect = false;
    static public Boolean fPrefRoundNearestTenth = false;
    static public AutoFillOptions fPrefAutoFillHobbs = AutoFillOptions.None;
    static public AutoFillOptions fPrefAutoFillTime = AutoFillOptions.None;
    static public NightCriteria NightPref = NightCriteria.EndOfCivilTwilight;
    static public NightLandingCriteria NightLandingPref = NightLandingCriteria.SunsetPlus60;

    public static Boolean IsRecording = false;
    public static Boolean IsFlying = false;
    public static Boolean HasPendingLanding = false;
    private Boolean IsListening = false;

    Boolean fNoRecord = false;

    // A single shared MFBLocation for the app
    private static MFBLocation m_Location = null;

    private BroadcastReceiver mReceiver = null;

    private void Init(Context c) {
        if (c != null) {
            if (mReceiver == null)
                mReceiver = new BroadcastReceiver() {
                    @Override
                    public void onReceive(Context context, Intent intent) {
                        Parcelable l = intent.getParcelableExtra(mfblocationservice.EXTRA_LOCATION);
                        if (l instanceof Location)
                            onLocationChanged((Location) l);
                    }
                };
        }
    }

    private Boolean fCheckPermissions(Context c) {
        return c != null && ContextCompat.checkSelfPermission(c, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;
    }

    public void startListening(Context c) {
        if (!IsListening && HasGPS(c) && !MFBConstants.fFakeGPS) {
            if (!fCheckPermissions(c))
                return;

            try {
                LocalBroadcastManager.getInstance(c).registerReceiver(mReceiver, new IntentFilter(mfblocationservice.ACTION_LOCATION_BROADCAST));

                Log.w(MFBConstants.LOG_TAG, String.format("Start Listening, Isrecording = %s", IsRecording ? "Yes" : "No"));

                // start background service
                // We only have 5 seconds from startForegroundService to calling startForeground, so let's build the Notification here first.
                Intent i = new Intent(c, mfblocationservice.class);
                if (Build.VERSION.SDK_INT >= 26)
                    c.startForegroundService(i);
                else
                    c.startService(i);

                IsListening = true;
            } catch (IllegalArgumentException ex) {
                MFBUtil.Alert(c, c.getString(R.string.errNoGPSTitle), c.getString(R.string.errCantUseGPS) + ex.getMessage());
            } catch (SecurityException ignored) {
            }
        }
    }

    public void stopListening(Context c) {
        if (IsListening && HasGPS(c)) {
            if (!fCheckPermissions(c))
                return;

            try {
                LocalBroadcastManager.getInstance(c).unregisterReceiver(mReceiver);
                c.stopService(new Intent(c, mfblocationservice.class));
                IsListening = false;
            } catch (SecurityException ignored) {
            }
        }
    }

    public MFBLocation(Context c, boolean fStartNow) {
        super();
        Init(c);
        if (fStartNow)
            startListening(c);
    }

    public void SetListener(FlightEvents l) {
        m_Listener = l;
        InformListenerOfStatus(m_lastSeenLoc);
    }

    public MFBLocation(Context c, FlightEvents l) {
        super();
        m_Listener = l;
        Init(c);
        startListening(c);
    }

    public Location CurrentLoc() {
        return m_lastSeenLoc;
    }

    public static Location LastSeenLoc() {
        return m_lastSeenLoc;
    }

    public static boolean HasGPS(Context c) {
        if (c == null)
            return false;

        try {
            LocationManager lm = (LocationManager) c.getSystemService(Context.LOCATION_SERVICE);
            assert lm != null;
            return lm.isProviderEnabled(LocationManager.GPS_PROVIDER);
        } catch (IllegalArgumentException ex) {
            MFBUtil.Alert(c, c.getString(R.string.errNoGPSTitle), ex.getMessage());
        }
        return false;
    }

    public static MFBLocation GetMainLocation() {
        return m_Location;
    }

    public static void setMainLocation(MFBLocation loc) {
        m_Location = loc;
    }

    public void ResetFlightData() {
        SQLiteDatabase db = MFBMain.mDBHelper.getWritableDatabase();
        try {
            db.delete("FlightTrack", null, null);
        } catch (Exception e) {
            Log.e(MFBConstants.LOG_TAG, "Unable to clear track in ResetFlightData");
        }
        IsRecording = false;
        IsFlying = false;
        HasPendingLanding = false;
        InformListenerOfStatus(m_lastSeenLoc);
    }

    public String getFlightDataString() {
        return LocSample.FlightDataFromSamples(LocSample.flightPathFromDB());
    }

    private void InformListenerOfStatus(Location newLoc) {
        if (m_Listener == null)
            return;

        if (newLoc == null) {
            m_Listener.UpdateStatus(GPSQuality.Unknown, IsFlying, null, IsRecording);
            return;
        }

        GPSQuality q = GPSQuality.Unknown;
        if (newLoc.hasAccuracy()) {
            float Accuracy = newLoc.getAccuracy();
            if (Accuracy <= 0.0 || Accuracy > MFBConstants.MIN_ACCURACY)
                q = GPSQuality.Poor;
            else
                q = (Accuracy < (MFBConstants.MIN_ACCURACY / 2.0)) ? GPSQuality.Excellent : GPSQuality.Good;
        }

        m_Listener.UpdateStatus(q, IsFlying, newLoc, IsRecording);
    }

    // determines if the sample is night for purposes of logging
    private boolean IsNightForFlight(SunriseSunsetTimes sst) {
        // short-circuit if we know it's day
        if (!sst.IsNight)
            return false;

        switch (NightPref) {
            case EndOfCivilTwilight:
                return sst.IsCivilNight;
            case Sunset:
                return true;    // since we short-circuited above, sst.IsNight must be true!
            case SunsetPlus15:
            case SunsetPlus30:
            case SunsetPlus60:
                return sst.IsWithinNightOffset;
        }
        return false;
    }

    private int NightFlightSunsetOffset() {
        switch (NightPref) {
            default:
            case Sunset:
            case EndOfCivilTwilight:
            case SunsetPlus60:
                return 60;
            case SunsetPlus15:
                return 15;
            case SunsetPlus30:
                return 30;
        }
    }

    public void onLocationChanged(@NonNull Location newLoc) {
        LocSample loc = new LocSample(newLoc);

        double SpeedKts = loc.Speed;
        double Accuracy = loc.HError;
        boolean fValidQuality = (Accuracy > 0 && Accuracy < MFBConstants.MIN_ACCURACY);
        boolean fValidSpeed = (SpeedKts > 0.1);
        boolean fValidTime = true;

        m_lastSeenLoc = newLoc; // always update this, even if we don't use it.

        if (m_currentLoc == null)
            m_currentLoc = m_lastSeenLoc;

        // see if things are too tightly spaced
        // get the time interval since the last location update
        long dt = loc.TimeStamp.getTime() - m_currentLoc.getTime();
        if (dt < 0) // new timestamp is prior to more recently seen one
            return;

        if ((IsFlying && dt < MFBConstants.MIN_SAMPLE_RATE_AIRBORNE) ||
                (!IsFlying && dt < MFBConstants.MIN_SAMPLE_RATE_TAXI))
            fValidTime = fPrefRecordFlightHighRes;   // i.e., false unless we are recording high res./

        if (!m_fIsEnabled)
            return;

        // only do detection/recording with quality samples
        if ((++cSamplesSinceWaking > MFBConstants.BOGUS_SAMPLE_COUNT) && fValidSpeed && fValidQuality && fValidTime) {
            SunriseSunsetTimes sst = new SunriseSunsetTimes(loc.TimeStamp, loc.Latitude, loc.Longitude, NightFlightSunsetOffset());

            // it's a good sample - use it
            m_currentLoc = newLoc;

            boolean fIsNightForFlight = IsNightForFlight(sst);
            boolean fIsNightForLandings = (NightLandingPref == NightLandingCriteria.Night) ? fIsNightForFlight : sst.IsFAANight;

            if (PreviousLoc != null && fPreviousLocWasNight && fIsNightForFlight && MFBLocation.fPrefAutoDetect) {
                double t = (newLoc.getTime() - PreviousLoc.getTime()) / 3600000.0;
                if (t < .5 && m_Listener != null)    // limit of half an hour between samples for night time
                    m_Listener.AddNightTime(t);
            }
            fPreviousLocWasNight =fIsNightForFlight;

            PreviousLoc = newLoc;

            // detect takeoffs/landings  These are different speeds to prevent bouncing
            if (MFBLocation.fPrefAutoDetect) {
                if (IsFlying) {
                    if (SpeedKts < MFBTakeoffSpeed.getLandingSpeed()) {
                        IsFlying = false;
                        loc.Comment = "Landing detected";
                        Log.w(MFBConstants.LOG_TAG, "Landing detected...");
                        HasPendingLanding = true;
                        if (m_Listener != null) {
                            Log.w(MFBConstants.LOG_TAG, "Notifying listener...");
                            m_Listener.LandingDetected(newLoc);
                        }
                    }
                } else {
                    if (SpeedKts > MFBTakeoffSpeed.getTakeOffspeed()) {
                        IsFlying = true;
                        loc.Comment = MFBMain.getResourceString(R.string.telemetryTakeOff);
                        HasPendingLanding = false; // back in the air - can't be a FS landing
                        if (m_Listener != null) {
                            m_Listener.TakeoffDetected(newLoc, fIsNightForLandings);
                        }
                    }
                }

                // see if we've had a full-stop landing
                if (SpeedKts < MFBConstants.FULL_STOP_SPEED && HasPendingLanding) {
                    if (m_Listener != null)
                        m_Listener.FSLandingDetected(fIsNightForLandings);
                    Log.w(MFBConstants.LOG_TAG, "FS " + (fIsNightForLandings ? "night " : "") + "landing detected");
                    loc.Comment = MFBMain.getResourceString(fIsNightForLandings ? R.string.telemetryFSNight : R.string.telemetryFSLanding);
                    HasPendingLanding = false;
                }
            }

            if (MFBLocation.fPrefRecordFlight && IsRecording && !ActNewFlight.fPaused && !fNoRecord) {
                SQLiteDatabase db = MFBMain.mDBHelper.getWritableDatabase();
                try {
                    ContentValues cv = new ContentValues();
                    loc.ToContentValues(cv);
                    long l = db.insert("FlightTrack", null, cv);
                    if (l < 0)
                        throw new Exception("Error saving to flight track");
                } catch (Exception e) {
                    Log.e(MFBConstants.LOG_TAG, "Unable to save to flight track");
                }
            }
        }

        // rate the quality
        InformListenerOfStatus(newLoc);
    }

    public void setIsRecording(Boolean f) {
        IsRecording = (fPrefRecordFlight && f);
    }

    public Boolean getIsRecording() {
        return (fPrefRecordFlight && IsRecording);
    }

    public void onProviderDisabled(@NonNull String provider) {
        m_fIsEnabled = false;
    }

    public void onProviderEnabled(@NonNull String provider) {
        m_fIsEnabled = true;
    }

    public void onStatusChanged(String provider, int status, Bundle extras) {
    }

}

File: app/src/main/java/model/MFBLocation.kt
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2025 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package model

import android.Manifest
import android.content.*
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Build
import android.os.Parcelable
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.myflightbook.android.ActNewFlight
import com.myflightbook.android.MFBMain
import com.myflightbook.android.MFBlocationservice
import com.myflightbook.android.R
import model.LocSample.Companion.flightDataFromSamples
import model.LocSample.Companion.flightPathFromDB

class MFBLocation : LocationListener {
    enum class GPSQuality {
        Unknown, Poor, Good, Excellent
    }

    interface FlightEvents {
        fun takeoffDetected(l: Location, fIsNight: Boolean)
        fun landingDetected(l: Location)
        fun landingDetectedFS(fIsNight: Boolean)
        fun addNightTime(t: Double)
        fun updateStatus(
            quality: GPSQuality,
            fAirborne: Boolean,
            loc: Location?,
            fRecording: Boolean
        )

        fun shouldKeepListening(): Boolean
    }

    private var mFisenabled = true
    private var mListener: FlightEvents? = null
    private var mCurrentloc: Location? = null // last good location we've seen
    private var cSamplesSinceWaking = 0
    private var fPreviousLocWasNight = false
    private var previousLoc: Location? = null

    enum class AutoFillOptions {
        None, FlightTime, EngineTime, HobbsTime, BlockTime, FlightStartToEngineEnd
    }

    enum class NightCriteria {
        EndOfCivilTwilight, Sunset, SunsetPlus15, SunsetPlus30, SunsetPlus60
    }

    enum class NightLandingCriteria {
        SunsetPlus60, Night
    }

    private var isListening = false
    var fNoRecord = false
    private var mReceiver: BroadcastReceiver? = null
    private fun init(c: Context?) {
        if (c != null) {
            if (mReceiver == null) mReceiver = object : BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {
                    val l = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
                        intent.getParcelableExtra(MFBlocationservice.EXTRA_LOCATION, Location::class.java)
                    else
                        @Suppress("DEPRECATION")
                        intent.getParcelableExtra<Parcelable>(MFBlocationservice.EXTRA_LOCATION)
                    if (l is Location) onLocationChanged(l)
                }
            }
        }
    }

    private fun fCheckPermissions(c: Context?): Boolean {
        return c != null && ContextCompat.checkSelfPermission(
            c,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
    }

    fun startListening(c: Context?) {
        if (!isListening && hasGPS(c) && !MFBConstants.FAKE_GPS) {
            if (c == null || !fCheckPermissions(c)) return
            try {
                LocalBroadcastManager.getInstance(c).registerReceiver(
                    mReceiver!!,
                    IntentFilter(MFBlocationservice.ACTION_LOCATION_BROADCAST)
                )
                Log.w(
                    MFBConstants.LOG_TAG,
                    String.format(
                        "Start Listening, Isrecording = %s",
                        if (IsRecording) "Yes" else "No"
                    )
                )

                // start background service
                // We only have 5 seconds from startForegroundService to calling startForeground, so let's build the Notification here first.
                val i = Intent(c, MFBlocationservice::class.java)
                c.startForegroundService(i)
                isListening = true
            } catch (ex: IllegalArgumentException) {
                MFBUtil.alert(
                    c,
                    c.getString(R.string.errNoGPSTitle),
                    c.getString(R.string.errCantUseGPS) + ex.message
                )
            } catch (ignored: SecurityException) {
            }
        }
    }

    fun stopListening(c: Context) {
        if (isListening && hasGPS(c)) {
            if (!fCheckPermissions(c)) return
            try {
                LocalBroadcastManager.getInstance(c).unregisterReceiver(mReceiver!!)
                c.stopService(Intent(c, MFBlocationservice::class.java))
                isListening = false
            } catch (ignored: SecurityException) {
            }
        }
    }

    constructor(c: Context, fStartNow: Boolean) : super() {
        init(c)
        if (fStartNow) startListening(c)
    }

    fun setListener(l: FlightEvents?) {
        mListener = l
        informListenerOfStatus(m_lastSeenLoc)
    }

    constructor(c: Context?, l: FlightEvents?) : super() {
        mListener = l
        init(c)
        startListening(c)
    }

    fun currentLoc(): Location? {
        return m_lastSeenLoc
    }

    fun resetFlightData() {
        val db = MFBMain.mDBHelper!!.writableDatabase
        try {
            db.delete("FlightTrack", null, null)
        } catch (e: Exception) {
            Log.e(MFBConstants.LOG_TAG, "Unable to clear track in ResetFlightData")
        }
        IsRecording = false
        IsFlying = false
        HasPendingLanding = false
        informListenerOfStatus(m_lastSeenLoc)
    }

    val flightDataString: String
        get() = flightDataFromSamples(flightPathFromDB())

    private fun informListenerOfStatus(newLoc: Location?) {
        if (mListener == null) return
        if (newLoc == null) {
            mListener!!.updateStatus(GPSQuality.Unknown, IsFlying, null, IsRecording)
            return
        }
        var q = GPSQuality.Unknown
        if (newLoc.hasAccuracy()) {
            val accuracy = newLoc.accuracy
            q =
                if (accuracy <= 0.0 || accuracy > MFBConstants.MIN_ACCURACY) GPSQuality.Poor else if (accuracy < MFBConstants.MIN_ACCURACY / 2.0) GPSQuality.Excellent else GPSQuality.Good
        }
        mListener!!.updateStatus(q, IsFlying, newLoc, IsRecording)
    }

    // determines if the sample is night for purposes of logging
    private fun isNightForFlight(sst: SunriseSunsetTimes): Boolean {
        // short-circuit if we know it's day
        return if (!sst.isNight!!) false else when (NightPref) {
            NightCriteria.EndOfCivilTwilight -> sst.isCivilNight!!
            NightCriteria.Sunset -> true // since we short-circuited above, sst.IsNight must be true!
            NightCriteria.SunsetPlus15, NightCriteria.SunsetPlus30, NightCriteria.SunsetPlus60 -> sst.isWithinNightOffset!!
        }
    }

    private fun nightFlightSunsetOffset(): Int {
        return when (NightPref) {
            NightCriteria.Sunset, NightCriteria.EndOfCivilTwilight, NightCriteria.SunsetPlus60 -> 60
            NightCriteria.SunsetPlus15 -> 15
            NightCriteria.SunsetPlus30 -> 30
        }
    }

    override fun onLocationChanged(newLoc: Location) {
        val loc = LocSample(newLoc)
        val speedKts = loc.speed
        val accuracy = loc.horizontalError
        val fValidQuality = accuracy > 0 && accuracy < MFBConstants.MIN_ACCURACY
        val fValidSpeed = speedKts > 0.1
        var fValidTime = true
        m_lastSeenLoc = newLoc // always update this, even if we don't use it.
        if (mCurrentloc == null) mCurrentloc = m_lastSeenLoc

        // see if things are too tightly spaced
        // get the time interval since the last location update
        val dt = loc.timeStamp.time - mCurrentloc!!.time
        if (dt < 0) // new timestamp is prior to more recently seen one
            return
        if (IsFlying && dt < MFBConstants.MIN_SAMPLE_RATE_AIRBORNE ||
            !IsFlying && dt < MFBConstants.MIN_SAMPLE_RATE_TAXI
        ) fValidTime = fPrefRecordFlightHighRes // i.e., false unless we are recording high res./
        if (!mFisenabled) return

        // only do detection/recording with quality samples
        if (++cSamplesSinceWaking > MFBConstants.BOGUS_SAMPLE_COUNT && fValidSpeed && fValidQuality && fValidTime) {
            val sst = SunriseSunsetTimes(
                loc.timeStamp,
                loc.latitude,
                loc.longitude,
                nightFlightSunsetOffset()
            )

            // it's a good sample - use it
            mCurrentloc = newLoc
            val fIsNightForFlight = isNightForFlight(sst)
            val fIsNightForLandings =
                if (NightLandingPref == NightLandingCriteria.Night) fIsNightForFlight else sst.isFAANight!!
            if (previousLoc != null && fPreviousLocWasNight && fIsNightForFlight && fPrefAutoDetect) {
                val t = (newLoc.time - previousLoc!!.time) / 3600000.0
                if (t < .5 && mListener != null) // limit of half an hour between samples for night time
                    mListener!!.addNightTime(t)
            }
            fPreviousLocWasNight = fIsNightForFlight
            previousLoc = newLoc

            // detect takeoffs/landings  These are different speeds to prevent bouncing
            if (fPrefAutoDetect) {
                if (IsFlying) {
                    if (speedKts < MFBTakeoffSpeed.landingSpeed) {
                        IsFlying = false
                        loc.comment = "Landing detected"
                        Log.w(MFBConstants.LOG_TAG, "Landing detected...")
                        HasPendingLanding = true
                        if (mListener != null) {
                            Log.w(MFBConstants.LOG_TAG, "Notifying listener...")
                            mListener!!.landingDetected(newLoc)
                        }
                    }
                } else {
                    if (speedKts > MFBTakeoffSpeed.takeOffspeed) {
                        IsFlying = true
                        loc.comment = MFBMain.getResourceString(R.string.telemetryTakeOff)
                        HasPendingLanding = false // back in the air - can't be a FS landing
                        if (mListener != null) {
                            mListener!!.takeoffDetected(newLoc, fIsNightForLandings)
                        }
                    }
                }

                // see if we've had a full-stop landing
                if (speedKts < MFBConstants.FULL_STOP_SPEED && HasPendingLanding) {
                    if (mListener != null) mListener!!.landingDetectedFS(fIsNightForLandings)
                    Log.w(
                        MFBConstants.LOG_TAG,
                        "FS " + (if (fIsNightForLandings) "night " else "") + "landing detected"
                    )
                    loc.comment =
                        MFBMain.getResourceString(if (fIsNightForLandings) R.string.telemetryFSNight else R.string.telemetryFSLanding)
                    HasPendingLanding = false
                }
            }
            if (fPrefRecordFlight && IsRecording && !ActNewFlight.fPaused && !fNoRecord) {
                val db = MFBMain.mDBHelper!!.writableDatabase
                try {
                    val cv = ContentValues()
                    loc.toContentValues(cv)
                    val l = db.insert("FlightTrack", null, cv)
                    if (l < 0) throw Exception("Error saving to flight track")
                } catch (e: Exception) {
                    Log.e(MFBConstants.LOG_TAG, "Unable to save to flight track")
                }
            }
        }

        // rate the quality
        informListenerOfStatus(newLoc)
    }

    var isRecording: Boolean
        get() = fPrefRecordFlight && IsRecording
        set(f) {
            IsRecording = fPrefRecordFlight && f
        }

    override fun onProviderDisabled(provider: String) {
        mFisenabled = false
    }

    override fun onProviderEnabled(provider: String) {
        mFisenabled = true
    }

    companion object {
        private var m_lastSeenLoc: Location? = null // last location we've seen
        @JvmField
        var fPrefRecordFlight = false
        @JvmField
        var fPrefRecordFlightHighRes = false
        @JvmField
        var fPrefAutoDetect = false
        @JvmField
        var fPrefRoundNearestTenth = false
        @JvmField
        var fPrefAutoFillHobbs = AutoFillOptions.None
        @JvmField
        var fPrefAutoFillTime = AutoFillOptions.None
        @JvmField
        var NightPref = NightCriteria.EndOfCivilTwilight
        @JvmField
        var NightLandingPref = NightLandingCriteria.SunsetPlus60
        @JvmField
        var IsRecording = false
        @JvmField
        var IsFlying = false
        @JvmField
        var HasPendingLanding = false

        // A single shared MFBLocation for the app
        private var m_Location: MFBLocation? = null
        fun lastSeenLoc(): Location? {
            return m_lastSeenLoc
        }

        @JvmStatic
        fun hasGPS(c: Context?): Boolean {
            if (c == null) return false
            try {
                val lm = (c.getSystemService(Context.LOCATION_SERVICE) as LocationManager)
                return lm.isProviderEnabled(LocationManager.GPS_PROVIDER)
            } catch (ex: IllegalArgumentException) {
                MFBUtil.alert(c, c.getString(R.string.errNoGPSTitle), ex.message)
            }
            return false
        }

        @JvmStatic
        fun getMainLocation(): MFBLocation? {
            return m_Location
        }

        @JvmStatic
        fun setMainLocation(loc: MFBLocation?) {
            m_Location = loc
        }
    }
}